src/systems/systems/src/electrical/emergency_generator.rs

```diff
@@ -1,63 +1,50 @@
-
-use crate::{
-    shared::{PowerConsumptionReport, RamAirTurbineHydraulicCircuitPressurised},
-    simulation::{InitContext, SimulationElement, SimulatorWriter, UpdateContext},
-};
+use crate::simulation::{InitContext, SimulationElement, SimulatorWriter, UpdateContext};
 
 use super::{
     ElectricalElement, ElectricalElementIdentifier, ElectricalElementIdentifierProvider,
     ElectricalStateWriter, ElectricitySource, Potential, PotentialOrigin, ProvideFrequency,
     ProvidePotential,
 };
+use crate::shared::{
+    EmergencyGeneratorPower, HydraulicGeneratorControlUnit, PowerConsumptionReport,
+};
+use uom::si::{
+    angular_velocity::revolution_per_minute, electric_potential::volt, f64::*, frequency::hertz,
+    power::watt,
+};
 
 pub struct EmergencyGenerator {
     identifier: ElectricalElementIdentifier,
     writer: ElectricalStateWriter,
     supplying: bool,
     output_frequency: Frequency,
     output_potential: ElectricPotential,
-    time_since_start: Duration,
-    starting_or_started: bool,
+    generated_power: Power,
+    demand: Power,
 }
 impl EmergencyGenerator {
+    const MIN_RPM_TO_SUPPLY_POWER: f64 = 10000.;
+    const MIN_POWER_TO_DECLARE_SUPPLYING_WATT: f64 = 100.;
+
     pub fn new(context: &mut InitContext) -> EmergencyGenerator {
         EmergencyGenerator {
             identifier: context.next_electrical_identifier(),
             writer: ElectricalStateWriter::new(context, "EMER_GEN"),
             supplying: false,
             output_frequency: Frequency::new::<hertz>(0.),
             output_potential: ElectricPotential::new::<volt>(0.),
-            time_since_start: Duration::from_secs(0),
-            starting_or_started: false,
-        }
-    }
-
-    pub fn update(
-        &mut self,
-        context: &UpdateContext,
-        hydraulic: &impl RamAirTurbineHydraulicCircuitPressurised,
-    ) {
-        // TODO: All of this is a very simple implementation.
-        // Once hydraulics is available we should improve it.
-        if self.starting_or_started {
-            self.time_since_start += context.delta();
+            generated_power: Power::new::<watt>(0.),
+            demand: Power::new::<watt>(0.),
         }
-
-        self.supplying = hydraulic.is_rat_hydraulic_circuit_pressurised()
-            && self.starting_or_started
-            && self.time_since_start > Duration::from_secs(8);
     }
 
-    pub fn start(&mut self) {
-        self.starting_or_started = true;
-    }
+    pub fn update(&mut self, gcu: &impl HydraulicGeneratorControlUnit) {
+        self.update_generated_power(gcu);
 
-    #[cfg(test)]
-    pub fn stop(&mut self) {
-        self.starting_or_started = false;
+        self.supplying = self.generated_power
+            > Power::new::<watt>(Self::MIN_POWER_TO_DECLARE_SUPPLYING_WATT)
+            || (gcu.motor_speed()
+                > AngularVelocity::new::<revolution_per_minute>(Self::MIN_RPM_TO_SUPPLY_POWER));
```

Comment:
**thought**: There might be automated tests missing for the expression to the right of `||`.